api/v1/server/handlers/workflows/cancel.go

```diff
@@ -0,0 +1,103 @@
+package workflows
+
+import (
+	"fmt"
+
+	"github.com/google/uuid"
+	"github.com/hashicorp/go-multierror"
+	"github.com/jackc/pgx/v5/pgtype"
+	"github.com/labstack/echo/v4"
+
+	"github.com/hatchet-dev/hatchet/api/v1/server/oas/gen"
+	"github.com/hatchet-dev/hatchet/api/v1/server/oas/transformers"
+	"github.com/hatchet-dev/hatchet/internal/msgqueue"
+	"github.com/hatchet-dev/hatchet/internal/repository/prisma/db"
+	"github.com/hatchet-dev/hatchet/internal/repository/prisma/dbsqlc"
+	"github.com/hatchet-dev/hatchet/internal/services/shared/tasktypes"
+)
+
+func (t *WorkflowService) WorkflowRunCancel(ctx echo.Context, request gen.WorkflowRunCancelRequestObject) (gen.WorkflowRunCancelResponseObject, error) {
+	tenant := ctx.Get("tenant").(*db.TenantModel)
+	runIds := request.Body.WorkflowRunIds
+
+	canceledStepRunsChannel := make(chan []*db.StepRunModel, len(runIds))
+	var returnErr error
+
+	for _, runId := range runIds {
+		go func(runId uuid.UUID) {
+			// Lookup step runs for the workflow run
+			runIdStr := runId.String()
+			stepRuns, err := t.config.EngineRepository.StepRun().ListStepRunsByWorkflowRunId(ctx.Request().Context(), tenant.ID, runIdStr)
+			if err != nil {
+				returnErr = multierror.Append(err, fmt.Errorf("failed to list step runs for workflow run %s", runIdStr))
+				canceledStepRunsChannel <- nil
+				return
+			}
+
+			var canceledStepRuns []*db.StepRunModel
+
+			for _, stepRun := range stepRuns {
+				status := stepRun.StepRun.Status
+
+				// If the step run is not in a final state, send a task to the taskqueue to cancel it
+				if status != dbsqlc.StepRunStatusSUCCEEDED && status != dbsqlc.StepRunStatusFAILED && status != dbsqlc.StepRunStatusCANCELLED {
+					var reason = "CANCELLED_BY_USER"
+					// send a task to the taskqueue
+					err = t.config.MessageQueue.AddMessage(
+						ctx.Request().Context(),
+						msgqueue.JOB_PROCESSING_QUEUE,
+						tasktypes.StepRunCancelToTask(stepRun, reason),
```

Comment:
I would modify the implementation of `JobRunCancelledToTask` to include a reason and cancel the job run with that task, instead of the step run. We don't want the API to be responsible for understanding the internals of looping over step runs and deciding which ones to cancel in a way that could eventually conflict with the jobs controller.

Reply:
not sure i'm following here, would we not still have to loop step runs to get the job ID?

Reply:
I'm basically suggesting to invoke this function instead: https://github.com/hatchet-dev/hatchet/blob/615ed4296b2d235975f4a48ceb0f1700b76ed6a0/internal/services/controllers/jobs/controller.go#L291

We would need to list job runs by a list of workflow run IDs to avoid an `n+1` query on the database.